Reorganize the repo and customize the build process

gfontenot opened the pull request.

gfontenot committed:
Don't build a standalone library

Building a library aside from our website wasn't really needed. We can
instead just build a single executable that includes all necessary
modules. This also keeps our build dependencies in a single place.

gfontenot committed:
Customize build process

The `site` binary is actually _super_ useful, and has some interesting
shit in it. Plus, this is a really good opportunity to learn more about
cabal.

This process is being cribbed from
https://github.com/blaenk/blaenk.github.io/blob/source/Setup.hs

I've renamed some things and broken some things up in an effort to grok
what was going on. Essentially:

 - We look at the build package description and make sure we have an
   executable named 'site'
 - if so, we get the build directory from the build info, and copy the
   'site' binary out of there, and into the pwd.

This will allow us to run `site` directly without having to go through
cabal or dig through the `dist` folder.

sidraval commented on Setup.hs:
```diff
@@ -0,0 +1,37 @@
+
+main :: IO ()
+main = do
+  defaultMainWithHooks $ simpleUserHooks { postBuild = copyBinary }
```
Do you need the `do` here?

gfontenot replied:
Sure don't!

sidraval commented on Setup.hs:
```diff
@@ -0,0 +1,37 @@
+  let exe = findExecutable "site" desc
+
+  when (isJust exe) $ do
+    let binaryName = fromJust exe
```
https://hackage.haskell.org/package/base-4.8.1.0/docs/src/Data.Maybe.html#fromJust

``` haskell
do
  exe <- findExecutable "site" doc
  buildDir buildInfo
  putStrLn $ "Copying executable '" ++ exe ++ "' to current directory..."
  copyFile (dir </> "site" </> binaryName) binaryName
```

That's kind of a shot in the dark, but something like that should work? Maybe with a `liftIO` or something?

gfontenot replied:
hmmmm, maybe a `guard` statement? https://hackage.haskell.org/package/base-4.8.1.0/docs/Control-Monad.html#v:guard

gfontenot replied:
I knew about `fromJust` being unsafe, but I kinda like the explicitness here about it being a noop if we don't have the executable.

Honestly, I've been thinking about this function since I pushed this, and it _could_ just be

``` haskell
do
    let exe = "site"
    buildDir buildInfo
    putStrLn $ "Copying executable '" ++ exe ++ "' to current directory..."
    copyFile (dir </> exe </> binaryName) binaryName
```

The whole `findExecutable` bit is probably overly safe.

sidraval replied:
Oh, yeah, go for that!

gfontenot committed:
Simplify setup

This whole thing was _way_ more generic than it needed to be, and super
defensive. This script isn't going to be re-used, and we know there is a
`site` binary. Lets act like it.

gfontenot committed:
Don't use unnecessary `do`, extract hooks

We didn't actually need to use `do` here. Also, it's a bit nicer to
extract the hooks to their own function.

gfontenot commented:
_so_ much cleaner
